File: ios/App/View/Blog/BlogView.swift
import SwiftUI

/// The blog's reading palette, mirrored from BlogPage.styles.css.ts.
private enum BlogPalette {
  static let paper = Color(red: 0.988, green: 0.984, blue: 0.965)
  static let ink = Color(red: 0.125, green: 0.129, blue: 0.110)
  static let inkSoft = Color(red: 0.443, green: 0.443, blue: 0.416)
  static let line = Color(red: 0.898, green: 0.890, blue: 0.847)
  static let accent = Color(red: 0.243, green: 0.420, blue: 0.310)
  static let accentSoft = Color(red: 0.933, green: 0.949, blue: 0.918)
  static let codeBg = Color(red: 0.949, green: 0.941, blue: 0.910)
}

struct BlogView: View {
  @State private var activeTag: String?
  @State private var openSlug: String?

  private var orderedPosts: [BlogContent.Post] { BlogContent.sortedPosts() }

  private var visiblePosts: [BlogContent.Post] {
    guard let tag = activeTag else { return orderedPosts }
    return orderedPosts.filter { $0.tags.contains(tag) }
  }

  var body: some View {
    ZStack {
      BlogPalette.paper.ignoresSafeArea()
      if let slug = openSlug, let post = orderedPosts.first(where: { $0.slug == slug }) {
        articleView(post)
      } else {
        listView
      }
    }
  }

  // MARK: - List

  private var listView: some View {
    ScrollView {
      VStack(alignment: .leading, spacing: 0) {
        masthead
        tagRow
          .padding(.vertical, 14)
        ForEach(visiblePosts) { post in
          postCard(post)
        }
        Text("\(BlogContent.title) — written by \(BlogContent.author.name). Built with Repobot.")
          .font(.footnote)
          .foregroundStyle(BlogPalette.inkSoft)
          .padding(.top, 28)
      }
      .padding(.horizontal, 22)
      .padding(.bottom, 48)
    }
  }

  private var masthead: some View {
    VStack(alignment: .leading, spacing: 10) {
      Text(BlogContent.title)
        .font(.system(size: 36, weight: .bold, design: .serif))
        .foregroundStyle(BlogPalette.ink)
        .padding(.top, 26)
      Text(BlogContent.tagline)
        .font(.subheadline)
        .foregroundStyle(BlogPalette.inkSoft)
      HStack(spacing: 12) {
        Text(BlogContent.author.initials)
          .font(.footnote.weight(.bold))
          .foregroundStyle(BlogPalette.paper)
          .frame(width: 40, height: 40)
          .background(BlogPalette.accent, in: Circle())
        VStack(alignment: .leading, spacing: 2) {
          Text(BlogContent.author.name)
            .font(.footnote.weight(.semibold))
            .foregroundStyle(BlogPalette.ink)
          Text(BlogContent.author.role)
            .font(.caption)
            .foregroundStyle(BlogPalette.inkSoft)
        }
      }
      .padding(.top, 6)
      Rectangle()
        .fill(BlogPalette.line)
        .frame(height: 1)
        .padding(.top, 16)
    }
  }

  private var tagRow: some View {
    ScrollView(.horizontal, showsIndicators: false) {
      HStack(spacing: 8) {
        tagChip(label: "All", isActive: activeTag == nil) { activeTag = nil }
        ForEach(BlogContent.allTags(in: orderedPosts), id: \.self) { tag in
          tagChip(label: tag, isActive: activeTag == tag) {
            activeTag = activeTag == tag ? nil : tag
          }
        }
      }
    }
  }

  private func tagChip(label: String, isActive: Bool, action: @escaping () -> Void) -> some View {
    Button(action: action) {
      Text(label)
        .font(.footnote.weight(.semibold))
        .foregroundStyle(isActive ? BlogPalette.paper : BlogPalette.inkSoft)
        .padding(.horizontal, 14)
        .padding(.vertical, 7)
        .background(
          Capsule().fill(isActive ? BlogPalette.accent : Color.clear))
        .overlay(
          Capsule().strokeBorder(isActive ? BlogPalette.accent : BlogPalette.line))
    }
    .buttonStyle(.plain)
  }

  private func postCard(_ post: BlogContent.Post) -> some View {
    Button {
      openSlug = post.slug
    } label: {
      VStack(alignment: .leading, spacing: 8) {
        postMeta(post, showTags: false)
        Text(post.title)
          .font(.system(size: 22, weight: .semibold, design: .serif))
          .foregroundStyle(BlogPalette.ink)
          .multilineTextAlignment(.leading)
        Text(post.summary)
          .font(.subheadline)
          .foregroundStyle(BlogPalette.inkSoft)
          .multilineTextAlignment(.leading)
        Rectangle()
          .fill(BlogPalette.line)
          .frame(height: 1)
          .padding(.top, 14)
      }
      .frame(maxWidth: .infinity, alignment: .leading)
      .padding(.top, 18)
    }
    .buttonStyle(.plain)
  }

  private func postMeta(_ post: BlogContent.Post, showTags: Bool) -> some View {
    HStack(spacing: 8) {
      Text(Self.formatDate(post.date))
      Text("·")
      Text("\(BlogMarkdown.readingTimeMinutes(post.body)) min read")
      if showTags {
        ForEach(post.tags, id: \.self) { tag in
          Text(tag)
            .font(.caption2.weight(.semibold))
            .foregroundStyle(BlogPalette.accent)
            .padding(.horizontal, 9)
            .padding(.vertical, 2)
            .background(Capsule().fill(BlogPalette.accentSoft))
        }
      }
    }
    .font(.caption)
    .foregroundStyle(BlogPalette.inkSoft)
  }

  // MARK: - Article

  private func articleView(_ post: BlogContent.Post) -> some View {
    ScrollView {
      VStack(alignment: .leading, spacing: 0) {
        Button {
          openSlug = nil
        } label: {
          Label("All posts", systemImage: "arrow.left")
            .font(.footnote.weight(.semibold))
            .foregroundStyle(BlogPalette.accent)
        }
        .buttonStyle(.plain)
        .padding(.top, 22)

        Text(post.title)
          .font(.system(size: 30, weight: .bold, design: .serif))
          .foregroundStyle(BlogPalette.ink)
          .padding(.top, 14)
        postMeta(post, showTags: true)
          .padding(.top, 8)

        VStack(alignment: .leading, spacing: 16) {
          ForEach(Array(BlogMarkdown.parseMarkdown(post.body).enumerated()), id: \.offset) {
            _, block in
            blockView(block)
          }
        }
        .padding(.top, 24)

        Rectangle()
          .fill(BlogPalette.line)
          .frame(height: 1)
          .padding(.vertical, 24)
        Text("\(BlogContent.author.name) — \(BlogContent.author.bio)")
          .font(.footnote)
          .foregroundStyle(BlogPalette.inkSoft)
      }
      .padding(.horizontal, 22)
      .padding(.bottom, 48)
    }
  }

  @ViewBuilder
  private func blockView(_ block: BlogBlock) -> some View {
    switch block {
    case .heading(let level, let inlines):
      Text(Self.attributed(inlines))
        .font(.system(size: level == 1 ? 26 : level == 2 ? 22 : 19, weight: .semibold, design: .serif))
        .foregroundStyle(BlogPalette.ink)
        .padding(.top, 12)
    case .paragraph(let inlines):
      Text(Self.attributed(inlines))
        .font(.body)
        .foregroundStyle(BlogPalette.ink)
        .lineSpacing(5)
    case .quote(let inlines):
      HStack(alignment: .top, spacing: 12) {
        RoundedRectangle(cornerRadius: 2)
          .fill(BlogPalette.accent)
          .frame(width: 3)
        Text(Self.attributed(inlines))
          .font(.body.italic())
          .foregroundStyle(BlogPalette.inkSoft)
          .lineSpacing(5)
      }
      .padding(.vertical, 4)
    case .code(_, let text):
      ScrollView(.horizontal, showsIndicators: false) {
        Text(text)
          .font(.system(size: 13, design: .monospaced))
          .foregroundStyle(BlogPalette.ink)
          .padding(14)
      }
      .background(
        RoundedRectangle(cornerRadius: 10)
          .fill(BlogPalette.codeBg))
      .overlay(
        RoundedRectangle(cornerRadius: 10)
          .strokeBorder(BlogPalette.line))
    case .list(let ordered, let items):
      VStack(alignment: .leading, spacing: 8) {
        ForEach(Array(items.enumerated()), id: \.offset) { index, item in
          HStack(alignment: .top, spacing: 10) {
            Text(ordered ? "\(index + 1)." : "•")
              .font(.body.weight(.semibold))
              .foregroundStyle(BlogPalette.accent)
            Text(Self.attributed(item))
              .font(.body)
              .foregroundStyle(BlogPalette.ink)
              .lineSpacing(4)
          }
        }
      }
    case .divider:
      Rectangle()
        .fill(BlogPalette.line)
        .frame(width: 96, height: 1)
        .frame(maxWidth: .infinity)
        .padding(.vertical, 8)
    }
  }

  /// Render the parsed inline marks into one AttributedString.
  private static func attributed(_ inlines: [BlogInline]) -> AttributedString {
    var result = AttributedString()
    for inline in inlines {
      switch inline {
      case .text(let s):
        result += AttributedString(s)
      case .bold(let s):
        var a = AttributedString(s)
        a.font = .body.weight(.bold)
        result += a
      case .italic(let s):
        var a = AttributedString(s)
        a.font = .body.italic()
        result += a
      case .code(let s):
        var a = AttributedString(s)
        a.font = .system(size: 15, design: .monospaced)
        a.backgroundColor = BlogPalette.codeBg
        result += a
      case .link(let text, let href):
        var a = AttributedString(text)
        a.link = URL(string: href)
        a.foregroundColor = BlogPalette.accent
        a.underlineStyle = .single
        result += a
      }
    }
    return result
  }

  private static func formatDate(_ iso: String) -> String {
    let parser = DateFormatter()
    parser.dateFormat = "yyyy-MM-dd"
    parser.locale = Locale(identifier: "en_US_POSIX")
    guard let date = parser.date(from: iso) else { return iso }
    let formatter = DateFormatter()
    formatter.dateStyle = .long
    formatter.locale = Locale(identifier: "en_US")
    return formatter.string(from: date)
  }
}

#Preview {
  BlogView()
}

File: android/app/src/main/kotlin/com/baseapp/android/view/blog/BlogView.kt
package com.baseapp.android.view.blog

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.text.SimpleDateFormat
import java.util.Locale

/** The blog's reading palette, mirrored from BlogPage.styles.css.ts. */
private object BlogPalette {
    val paper = Color(0xFFFCFBF6)
    val ink = Color(0xFF20211C)
    val inkSoft = Color(0xFF71716A)
    val line = Color(0xFFE5E3D8)
    val accent = Color(0xFF3E6B4F)
    val accentSoft = Color(0xFFEEF2EA)
    val codeBg = Color(0xFFF2F0E8)
}

@Composable
fun BlogView() {
    var activeTag by remember { mutableStateOf<String?>(null) }
    var openSlug by remember { mutableStateOf<String?>(null) }

    val orderedPosts = remember { BlogContent.sortedPosts() }
    val openPost = openSlug?.let { slug -> orderedPosts.firstOrNull { it.slug == slug } }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(BlogPalette.paper),
    ) {
        if (openPost != null) {
            ArticleView(post = openPost, onBack = { openSlug = null })
        } else {
            PostListView(
                posts = orderedPosts,
                activeTag = activeTag,
                onTagChange = { activeTag = it },
                onOpen = { openSlug = it },
            )
        }
    }
}

@Composable
private fun PostListView(
    posts: List<BlogContent.Post>,
    activeTag: String?,
    onTagChange: (String?) -> Unit,
    onOpen: (String) -> Unit,
) {
    val visible = if (activeTag == null) posts else posts.filter { activeTag in it.tags }
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 22.dp)
            .padding(bottom = 48.dp),
    ) {
        Masthead()
        TagRow(posts = posts, activeTag = activeTag, onTagChange = onTagChange)
        visible.forEach { post -> PostCard(post = post, onOpen = onOpen) }
        Text(
            text = "${BlogContent.TITLE} — written by ${BlogContent.author.name}. Built with Repobot.",
            fontSize = 13.sp,
            color = BlogPalette.inkSoft,
            modifier = Modifier.padding(top = 28.dp),
        )
    }
}

@Composable
private fun Masthead() {
    Column(modifier = Modifier.padding(top = 40.dp)) {
        Text(
            text = BlogContent.TITLE,
            fontSize = 36.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Serif,
            color = BlogPalette.ink,
        )
        Text(
            text = BlogContent.TAGLINE,
            fontSize = 15.sp,
            lineHeight = 22.sp,
            color = BlogPalette.inkSoft,
            modifier = Modifier.padding(top = 8.dp),
        )
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(top = 18.dp),
        ) {
            Box(
                contentAlignment = Alignment.Center,
                modifier = Modifier
                    .size(40.dp)
                    .clip(CircleShape)
                    .background(BlogPalette.accent),
            ) {
                Text(
                    text = BlogContent.author.initials,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold,
                    color = BlogPalette.paper,
                )
            }
            Column(modifier = Modifier.padding(start = 12.dp)) {
                Text(
                    text = BlogContent.author.name,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = BlogPalette.ink,
                )
                Text(
                    text = BlogContent.author.role,
                    fontSize = 12.sp,
                    color = BlogPalette.inkSoft,
                )
            }
        }
        HorizontalRule(modifier = Modifier.padding(top = 18.dp))
    }
}

@Composable
private fun TagRow(
    posts: List<BlogContent.Post>,
    activeTag: String?,
    onTagChange: (String?) -> Unit,
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier
            .horizontalScroll(rememberScrollState())
            .padding(vertical = 14.dp),
    ) {
        TagChip(label = "All", isActive = activeTag == null) { onTagChange(null) }
        BlogContent.allTags(posts).forEach { tag ->
            TagChip(label = tag, isActive = activeTag == tag) {
                onTagChange(if (activeTag == tag) null else tag)
            }
        }
    }
}

@Composable
private fun TagChip(label: String, isActive: Boolean, onClick: () -> Unit) {
    Text(
        text = label,
        fontSize = 13.sp,
        fontWeight = FontWeight.SemiBold,
        color = if (isActive) BlogPalette.paper else BlogPalette.inkSoft,
        modifier = Modifier
            .clip(RoundedCornerShape(999.dp))
            .background(if (isActive) BlogPalette.accent else Color.Transparent)
            .border(
                width = 1.dp,
                color = if (isActive) BlogPalette.accent else BlogPalette.line,
                shape = RoundedCornerShape(999.dp),
            )
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp, vertical = 7.dp),
    )
}

@Composable
private fun PostCard(post: BlogContent.Post, onOpen: (String) -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onOpen(post.slug) }
            .padding(top = 18.dp),
    ) {
        PostMeta(post = post, showTags = false)
        Text(
            text = post.title,
            fontSize = 22.sp,
            fontWeight = FontWeight.SemiBold,
            fontFamily = FontFamily.Serif,
            color = BlogPalette.ink,
            modifier = Modifier.padding(top = 8.dp),
        )
        Text(
            text = post.summary,
            fontSize = 14.sp,
            lineHeight = 21.sp,
            color = BlogPalette.inkSoft,
            modifier = Modifier.padding(top = 6.dp),
        )
        HorizontalRule(modifier = Modifier.padding(top = 16.dp))
    }
}

@Composable
private fun PostMeta(post: BlogContent.Post, showTags: Boolean) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(
            text = "${formatDate(post.date)} · ${BlogMarkdown.readingTimeMinutes(post.body)} min read",
            fontSize = 12.sp,
            color = BlogPalette.inkSoft,
        )
        if (showTags) {
            post.tags.forEach { tag ->
                Text(
                    text = tag,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = BlogPalette.accent,
                    modifier = Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(BlogPalette.accentSoft)
                        .padding(horizontal = 9.dp, vertical = 2.dp),
                )
            }
        }
    }
}

@Composable
private fun ArticleView(post: BlogContent.Post, onBack: () -> Unit) {
    val blocks = remember(post.slug) { BlogMarkdown.parseMarkdown(post.body) }
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 22.dp)
            .padding(bottom = 48.dp),
    ) {
        Text(
            text = "← All posts",
            fontSize = 13.sp,
            fontWeight = FontWeight.SemiBold,
            color = BlogPalette.accent,
            modifier = Modifier
                .padding(top = 40.dp)
                .clickable(onClick = onBack),
        )
        Text(
            text = post.title,
            fontSize = 30.sp,
            lineHeight = 36.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Serif,
            color = BlogPalette.ink,
            modifier = Modifier.padding(top = 14.dp),
        )
        Box(modifier = Modifier.padding(top = 8.dp)) {
            PostMeta(post = post, showTags = true)
        }
        Column(
            verticalArrangement = Arrangement.spacedBy(16.dp),
            modifier = Modifier.padding(top = 24.dp),
        ) {
            blocks.forEach { block -> BlockView(block) }
        }
        HorizontalRule(modifier = Modifier.padding(vertical = 24.dp))
        Text(
            text = "${BlogContent.author.name} — ${BlogContent.author.bio}",
            fontSize = 13.sp,
            lineHeight = 19.sp,
            color = BlogPalette.inkSoft,
        )
    }
}

@Composable
private fun BlockView(block: BlogBlock) {
    when (block) {
        is BlogBlock.Heading -> Text(
            text = inlineAnnotated(block.inlines),
            fontSize = when (block.level) {
                1 -> 26.sp
                2 -> 22.sp
                else -> 19.sp
            },
            fontWeight = FontWeight.SemiBold,
            fontFamily = FontFamily.Serif,
            color = BlogPalette.ink,
            modifier = Modifier.padding(top = 12.dp),
        )
        is BlogBlock.Paragraph -> InlineText(block.inlines)
        is BlogBlock.Quote -> Row(modifier = Modifier.height(IntrinsicSize.Min)) {
            Box(
                modifier = Modifier
                    .width(3.dp)
                    .fillMaxHeight()
                    .clip(RoundedCornerShape(2.dp))
                    .background(BlogPalette.accent),
            )
            Text(
                text = inlineAnnotated(block.inlines),
                fontSize = 16.sp,
                lineHeight = 25.sp,
                fontStyle = FontStyle.Italic,
                color = BlogPalette.inkSoft,
                modifier = Modifier.padding(start = 12.dp),
            )
        }
        is BlogBlock.Code -> Text(
            text = block.text,
            fontSize = 12.5.sp,
            lineHeight = 19.sp,
            fontFamily = FontFamily.Monospace,
            color = BlogPalette.ink,
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(10.dp))
                .background(BlogPalette.codeBg)
                .border(1.dp, BlogPalette.line, RoundedCornerShape(10.dp))
                .horizontalScroll(rememberScrollState())
                .padding(14.dp),
        )
        is BlogBlock.ListBlock -> Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            block.items.forEachIndexed { index, item ->
                Row {
                    Text(
                        text = if (block.ordered) "${index + 1}." else "•",
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = BlogPalette.accent,
                    )
                    Spacer(modifier = Modifier.width(10.dp))
                    InlineText(item)
                }
            }
        }
        BlogBlock.Divider -> Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 8.dp),
        ) {
            Box(
                modifier = Modifier
                    .width(96.dp)
                    .height(1.dp)
                    .background(BlogPalette.line),
            )
        }
    }
}

private const val LINK_TAG = "URL"

@Composable
private fun InlineText(inlines: List<BlogInline>) {
    val uriHandler = LocalUriHandler.current
    val annotated = inlineAnnotated(inlines)
    val hasLink = annotated.getStringAnnotations(LINK_TAG, 0, annotated.length).isNotEmpty()
    if (hasLink) {
        androidx.compose.foundation.text.ClickableText(
            text = annotated,
            style = androidx.compose.ui.text.TextStyle(
                fontSize = 16.sp,
                lineHeight = 26.sp,
                color = BlogPalette.ink,
            ),
            onClick = { offset ->
                annotated.getStringAnnotations(LINK_TAG, offset, offset)
                    .firstOrNull()
                    ?.let { uriHandler.openUri(it.item) }
            },
        )
    } else {
        Text(
            text = annotated,
            fontSize = 16.sp,
            lineHeight = 26.sp,
            color = BlogPalette.ink,
        )
    }
}

private fun inlineAnnotated(inlines: List<BlogInline>): AnnotatedString =
    buildAnnotatedString {
        inlines.forEach { inline ->
            when (inline) {
                is BlogInline.Text -> append(inline.text)
                is BlogInline.Bold -> {
                    pushStyle(SpanStyle(fontWeight = FontWeight.Bold))
                    append(inline.text)
                    pop()
                }
                is BlogInline.Italic -> {
                    pushStyle(SpanStyle(fontStyle = FontStyle.Italic))
                    append(inline.text)
                    pop()
                }
                is BlogInline.Code -> {
                    pushStyle(
                        SpanStyle(
                            fontFamily = FontFamily.Monospace,
                            fontSize = 14.sp,
                            background = BlogPalette.codeBg,
                        ),
                    )
                    append(inline.text)
                    pop()
                }
                is BlogInline.Link -> {
                    pushStringAnnotation(tag = LINK_TAG, annotation = inline.href)
                    pushStyle(
                        SpanStyle(
                            color = BlogPalette.accent,
                            fontWeight = FontWeight.SemiBold,
                            textDecoration = TextDecoration.Underline,
                        ),
                    )
                    append(inline.text)
                    pop()
                    pop()
                }
            }
        }
    }

@Composable
private fun HorizontalRule(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(1.dp)
            .background(BlogPalette.line),
    )
}

private fun formatDate(iso: String): String {
    val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
    val date = parser.parse(iso) ?: return iso
    return SimpleDateFormat("MMMM d, yyyy", Locale.US).format(date)
}
